src/lowlevel/mpc.jl

```diff
@@ -0,0 +1,315 @@
+    @constraint(model, fill(λ, N) + D' * velocity_local >= zeros(N))  # (8)
+
+    λ_indicator = Variable(model)
+    @constraint(model, λ_indicator ∈ {0, 1})
+    # Equation (12)
+    # λ_indicator == 1 => μc_n - sum(β) == 0
+    # λ_indicator == 0 => λ == 0
+    c_n = @expression(dot(contact_point.force_local, FreeVector3D(contact_point.normal_aligned_frame, 0.0, 0.0, 1.0)))
+    @constraint(model, [μ * c_n - ones(N)' * β] <= [contact_model.boolean_params.ρ_max * (1 - λ_indicator)])
+    @constraint(model, [λ] <= [λ_indicator * contact_model.velocity_max])
+
+    β_indicator = [Variable(model) for _ in 1:N]
+    for i in 1:N
+        @constraint(model, β_indicator[i] ∈ {0, 1})
+    end
+    # Equation (11)
+    # β_indicator[i] == 1 => λ_plus_D'v[i] == 0
+    # β_indicator[i] == 0 => β[i] == 0
+    @constraint(model, fill(λ, N) + D' * velocity_local <= contact_model.velocity_max * (ones(N) - β_indicator))
+    @constraint(model, β <= contact_model.boolean_params.ρ_max * β_indicator)
+end
+
+"""
+Holds a few useful dynamics paramters which are re-used for each stage in the MPC
+optmization.
+"""
+struct DynamicsParams{TH, Tc, TJ}
+    H::TH
+    c::Tc
+    J_qv::TJ
+end
+
+function generalized_torque(state, contact_point::ContactPoint, model::Model)
+    mechanism = state.mechanism
+    body = body_fixed_frame_to_body(mechanism, contact_point.position.frame)
+    path_to_root = path(mechanism, body, root_body(mechanism))
+    J_body_to_root = let state = state, path_to_root = path_to_root
+        Parameter(geometric_jacobian(state, path_to_root), model) do J
+            geometric_jacobian!(J, state, path_to_root)
+        end
+    end
+    @expression adjoint(angular(J_body_to_root)) * angular(contact_point.wrench_world) + adjoint(linear(J_body_to_root)) * linear(contact_point.wrench_world)
+end
+
+function initialize!(stage::MPCStage, model::Model, state::MechanismState,
+                     params::DynamicsParams, q_prev, v_prev)
+    @assert !stage.initialized[]
+    Δt = stage.Δt
+    H = params.H
+    c = params.c
+    J_qv = params.J_qv
+    τ_ext = zeros(num_velocities(state))
+    for contact in stage.contacts
+        τ_ext = @expression τ_ext + generalized_torque(state, contact, model)
+    end
+    @constraint(model, H * (stage.v - v_prev) == Δt * (stage.u - c - τ_ext))
+    q̇ = @expression(J_qv * stage.v)
+    @constraint(model, q̇ == (1 / Δt) * (stage.q - q_prev))
+    @constraint(model, stage.v̇ == (1 / Δt) * (stage.v - v_prev))
+    stage.initialized[] = true
+end
+
+
+function initialize!(controller::MPCController)
+    @assert !controller.initialized[]
+    model = controller.qpmodel
+    # terminal_state_cost = controller.terminal_state_cost
+
+    H = let state = controller.state
+        Parameter(mass_matrix(state), model) do H
+            mass_matrix!(H, state)
+        end
+    end
+    c = let state = controller.state, result = controller.dynamicsresult
+        Parameter(result.dynamicsbias, model) do c
```

Comment:
A little weird that you don't use `c` here, but I guess it works.

Reply:
Ah, yeah, I'll make that inner `c` a `_`